xapian-bindings/python3/Makefile.am

```diff
@@ -172,3 +174,110 @@ $(sphinxdocs): xapian/__init__.py xapian/_xapian$(PYTHON3_SO) docs/conf.py $(RST
+_xapianletor_la_LIBADD = $(XAPIANLETOR_LIBS) $(PYTHON3_LIBS)
+
+# We want `xapianletor/__init__.py` rather than xapianletor.py so the module is treated
+# as a package.
+xapianletor/__init__.py: xapian/__init__.py xapianletor.py
+	$(MKDIR_P) xapianletor
+	cp `test -f xapianletor.py || echo '$(srcdir)/'`xapianletor.py xapianletor/__init__.py
+
+# We "import _xapianletor" first so that if we fail to import the glue library
+# we don't generate a broken .pyc or .pyo, and we do it with the current
+# directory set to where the module is as Python with -c always searches the
+# current directory first.
+#
+# We then "import xapianletor" as a separate command, as that's more like how things
+# are once installed - in particular this will fail if the SWIG shadow wrapper
+# doesn't cope with being xapianletor/__init__.py (currently it needs fixing
+# up, which is done by fixup-swig-py3-wrapper).
+xapianletor/__pycache__/__init__.@PYTHON3_CACHE_TAG@.pyc: xapian/__pycache__/__init__.@PYTHON3_CACHE_TAG@.pyc xapianletor/__init__.py xapianletor/_xapianletor$(PYTHON3_SO)
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -c 'import os;os.chdir("xapianletor");import _xapianletor'
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -c 'import xapianletor'
+
+xapianletor/__pycache__/__init__.@PYTHON3_CACHE_OPT1_EXT@: xapian/__pycache__/__init__.@PYTHON3_CACHE_OPT1_EXT@ xapianletor/__init__.py xapianletor/_xapianletor$(PYTHON3_SO)
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -O -c 'import os;os.chdir("xapianletor");import _xapianletor'
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -O -c 'import xapianletor'
+
+xapianletor/_xapianletor$(PYTHON3_SO): _xapian.la _xapianletor.la
+	$(MKDIR_P) xapianletor
+	$(LIBTOOL) --config > libtoolconfig.tmp
+## ksh requires a path on the sourced file.
+	. ./libtoolconfig.tmp; cp $$objdir/_xapianletor$(PYTHON3_SO) xapianletor
+	rm -f libtoolconfig.tmp
+
+CLEANFILES += \
+    xapianletor/_xapianletor$(PYTHON3_SO) \
+    xapianletor/__init__.py \
+    xapianletor/__pycache__/__init__.@PYTHON3_CACHE_TAG@.pyc \
+    xapianletor/__pycache__/__init__.@PYTHON3_CACHE_OPT1_EXT@ \
+    err_output_ListNet_2.txt \
+    ndcg_output_ListNet_2.txt \
+    training_output1.txt \
+    $(srcdir)/*.pyc
+
+if MAINTAINER_MODE
+BUILT_SOURCES += exceptletor.i
+exceptletor.i: $(srcdir)/generate-letor-python-exceptions ../../xapian-letor/exception_data.pm
+	$(PERL) -w -I$(srcdir)/../../xapian-letor $(srcdir)/generate-letor-python-exceptions
+
+stampletor = xapianletor_wrap.stamp
+RUN_SWIG_LETOR = stampletor='$(stampletor)' $(PERL) '$(top_srcdir)'/swig-depcomp $(SWIG)
+
+xapianletor_wrap.cc xapianletor_wrap.h xapianletor.py xapianletor_wrap.d: $(stampletor)
+	$(make_many_locked)
+$(stampletor): exceptletor.i fixup-swig-py3-wrapper
+	$(multitarget_begin)
+	$(RUN_SWIG_LETOR) $(SWIG_WERROR) -I. -I'$(srcdir)' $(SWIG_FLAGS) $(SWIG_LETOR_FLAGS) -c++ \
+	    -python -threads -shadow -O \
+	    -o xapianletor_wrap.cc '$(srcdir)/'pythonletor.i
+	$(PERL) -i $(srcdir)/fixup-swig-py3-wrapper xapianletor.py
+	$(multitarget_end)
```

Comment:
We ship the SWIG-generated sources, so these rules need to have run before `make dist` works.  That means `make dist` requires configure to have been run with `--enable-letor`.  If it wasn't, these files won't be found, and the rules to create them are disabled.  Or worse, if stale versions of them are present in the tree from a previous build with `--enable-letor` then those could get shipped.

I'm not quite sure how we want this to work, but the current way it does seems a bit problematic.

Not shipping the SWIG-generated sources is one answer, but because we build bindings for N languages we often have quite a tight SWIG version dependency because there's a recent fix for some language or other we really need.  So that would require users to install a particular SWIG build, and we'd probably end up fending off bug reports from users who don't read the instructions.

Reply:
Understood but It's not exactly clear what you want me to do here.

Reply:
I'm not quite sure exactly what we want to do here either.

I think the underlying problem is that conditionals in `Makefile.am` can interact with `make dist`, but the distribution tarballs really need to be reproducible, and so should not depend on configure options, etc.

I think automake can "see through" conditionals when determining the sources files to ship, so at least we wouldn't just end up creating a distribution which is missing the letor sources entirely, but `make` needs to be able to see rules to create files we should be shipping and it can't if those rules are inside conditionals, and `automake` can't really solve that.

So perhaps the rule to run swig should be outside the `if LETOR_FOUND` conditional, and similarly for any other rules which generate files to ship.  Just building the non-letor bindings should not trigger such rules, but at least then `make dist` can't end up shipping stale generated files.

This is all a bit fiddly to get right though.

Reply:
I think the most important thing will be to ensure that any `make dist` that _we_ (as the Xapian project) might do is run with `--enable-letor`. Putting a warning somewhere about this (`INSTALL`, perhaps?) is probably a good idea. Or even, if we can, spitting out a warning on `make dist` without `--enable-letor` in the configure run.

Reply:
Done.

Reply:
I tested building snapshot/release tarballs - currently this is the command that we build tarballs with:

```
env LD_LIBRARY_PATH="`pwd`"/xapian-core/.libs make -s distcheck VALGRIND=
```

And the obvious extension to that is:

```
env LD_LIBRARY_PATH="`pwd`"/xapian-core/.libs:"`pwd`"/xapian-letor/.libs make -s distcheck VALGRIND=
```

But this doesn't work because:

```
make[5]: *** No rule to make target 'xapianletor_wrap.cc', needed by 'distdir-am'.  Stop.
```

The error here seems to indicate you haven't implemented what I suggested above - and indeed looking at `python3/Makefile.am` the rules to generate these files are still conditional on `BUILD_LETOR_BINDINGS`.  Changing that is not enough to make this actually work (as it needs to find the right xapian-letor headers but `xapianletor-config` won't be consulted without `--enable-letor`), but as I noted before this change would at least avoid the risk of shipping stale generated files.

Just documenting that you need to do X for Y to work properly is all very well, but do you always read all the documentation for something before using it?  I know I don't, and I doubt most people do either.  If the result of not doing X might be to silently do the wrong thing that's likely to end up biting someone sooner or later, especially for something like `make dist` where no special steps are usually needed.

As for actually making the snapshot and release tarball building process work, I think we probably need to add `--enable-letor` to `DISTCHECK_CONFIGURE_FLAGS` - I tested this patch and it seems to work:

```
diff --git a/xapian-bindings/Makefile.am b/xapian-bindings/Makefile.am
index 45a0e3cb1fcb..d3a9ce41d1de 100644
--- a/xapian-bindings/Makefile.am
+++ b/xapian-bindings/Makefile.am
@@ -28,7 +28,8 @@ doxygen-xml/index.xml: $(XAPIAN_HEADERS)
 endif
 endif
 
-DISTCHECK_CONFIGURE_FLAGS = XAPIAN_CONFIG="$(XAPIAN_CONFIG)" \
+DISTCHECK_CONFIGURE_FLAGS = --enable-letor \
+	XAPIAN_CONFIG="$(XAPIAN_CONFIG)" \
 	XAPIANLETOR_CONFIG="$(XAPIANLETOR_CONFIG)" \
 	PYTHON2_LIB="`pwd`/../_inst/tmp_py2lib" \
 	PYTHON3_LIB="`pwd`/../_inst/tmp_py3lib" \
```

That's kind of ugly, but I don't really see a better approach.